### DISCUSSION TITLE
how to match all function with name in javascript ?

### DISCUSSION THREAD

**Discussion by WingDust** [ast-grep/ast-grep#716] [Q&A] [answered]
```sh
sg  -p 'function $FUNC($$$ARGS)' -l js                                                                                                               
Warning: Pattern contains an ERROR node and may cause unexpected results.
Help: ast-grep parsed the pattern but it matched nothing in this run. Try using playground to refine the pattern.
See also: https://ast-grep.github.io/playground.html

sg -V                                                                                                                                              
ast-grep 0.13.0
```
any syntax error ?

by the way , how to  ignore with `.gitignore` when grep in project

**Comment by HerringtonDarkholme** [comment, marked as answer]
You can see in the playground link how the pattern is parsed, as shown in the error message.

https://ast-grep.github.io/playground.html#eyJtb2RlIjoiUGF0Y2giLCJsYW5nIjoiamF2YXNjcmlwdCIsInF1ZXJ5IjoiZnVuY3Rpb24gJEZVTkMoJCQkQVJHUykiLCJyZXdyaXRlIjoiIiwiY29uZmlnIjoicnVsZTpcbiAgcGF0dGVybjogaWYgKCRBKSB7ICQkJEIgfVxuICBmb2xsb3dzOlxuICAgIHBhdHRlcm46IGlmICgkQSkgeyAkJCRDIH1cbiAgICBzdG9wQnk6IGVuZFxuZml4OiB8LVxuICBpZiAoJEEpIHsgXG4gICAgJCQkQ1xuICAgICQkJEJcbiAgfVxuLS0tXG5ydWxlOlxuICBwYXR0ZXJuOiBpZiAoJEEpIHsgJCQkQiB9XG4gIHByZWNlZGVzOlxuICAgIHBhdHRlcm46IGlmICgkQSkgeyAkJCRDIH1cbiAgICBzdG9wQnk6IGVuZFxuZml4OiAnJyIsInNvdXJjZSI6IiJ9


<img width="400" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/2921bb08-d6d4-42a5-b594-a32e571e6e28">

This is a valid pattern instead

<img width="434" alt="image" src="https://github.com/ast-grep/ast-grep/assets/2883231/cd4d0573-f4d9-45d8-a8b4-0ede8814f0e9">

**Comment by HerringtonDarkholme** [comment]
By the way, you can check out ast-grep's command line reference in the documentation site.

https://ast-grep.github.io/reference/cli.html#sg-run

https://ast-grep.github.io/reference/cli/run.html#no-ignore-file-type

You can also view the documentation on the command line by 

```bash
sg -h
sg --help
sg run -h
sg run --help
```

Either one will work.
